### PR TITLE
Position Multi-QMs

### PR DESCRIPTION
**Pull request by huyifanx**
This branch is based on Diallo's dynamic update branch.

### CODE REVIEW

File: Assets/Scripts/Rendering/State.cs [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
+    private const float UPDATE_INTERVAL = 10.0f;
     private float updateCountdown = UPDATE_INTERVAL;
 
     // Main dictionary keeping all connection and their rendered counterparts
     private Dictionary<MQ.Client, GameObject> qmgrs = new Dictionary<MQ.Client, GameObject>();
 
+    // Counter for recording the number of rendered QM
+    private int renderedQueueManagers = 0;
+    // Integer array for recording the postion of the last rendered manager
+    private List<int[]> lastQMPosition = new List<int[]>();
```
Comment (phillip-tian) [inline Assets/Scripts/Rendering/State.cs]: This list seems to contain all the rendered QM's positions, so maybe it should be called qmPositions?

File: Assets/Scripts/Rendering/State.cs [hunk 2]
```diff
@@ -207,5 +221,96 @@ public MQ.Message GetMessage(string selectedQMName, string selectedQueueName, st
...
+        int[] result = new int[] {0,0,0};
+        if (renderedQueueManagers==0)
+        {
+            lastQMPosition.Add(result);
+            return result;
+        }
+        if ((renderedQueueManagers+1)%2==0)
+        {
+            int[] lastQMSize = GetQueueManagerSize(qmgrsList[renderedQueueManagers-1]);
+            result[0]=lastQMPosition[renderedQueueManagers-1][0]+lastQMSize[0]+10;
```
Comment (phillip-tian) [inline Assets/Scripts/Rendering/State.cs]: Try not to use magic numbers (10).

File: Assets/Scripts/Rendering/State.cs [hunk 3]
```diff
@@ -207,5 +221,96 @@ public MQ.Message GetMessage(string selectedQMName, string selectedQueueName, st
...
+            result[1]=lastQMPosition[renderedQueueManagers-1][1];
+            result[2]=lastQMPosition[renderedQueueManagers-1][2];
+        }else{
+            int[] lastQMSize = GetQueueManagerSize(qmgrsList[renderedQueueManagers-2]);
+            result[0]=lastQMPosition[renderedQueueManagers-2][0];
+            result[1]=lastQMPosition[renderedQueueManagers-2][1];
+            result[2]=lastQMPosition[renderedQueueManagers-2][2]+lastQMSize[1]+10;
```
Comment (phillip-tian) [inline Assets/Scripts/Rendering/State.cs]: Magic number

**Comment by phillip-tian** [comment]
@lukasotocerny The code looks ok, I've pulled this branch and tested it locally and can confirm it correctly displays two QMs. I'm leaving this PR open for you to take a look at.
@huyifanx I made minor comments should you wish to clean the code up a little. Also, have you tried connecting to 3 QMs?

**Comment by huyifanx** [comment]
@phillip-tian Yes I have tried connecting to 3 QMs, It works well. And the three comments above are very reasonable, I will fix them

**Review by lukasotocerny — CHANGES_REQUESTED** [review]
Great that the functionality works! I have several remarks on how to improve the code so that we remove code duplication and make it cleaner. @huyifanx let's jump on a call today and have a pair coding session where we go through it together. @phillip-tian feel free to join as well. In the meanwhile please resolve merge conflicts as I merged Diallo's branch into main earlier today.

**Review by huyifanx — COMMENTED** [review]
Nice, it looks better now

**Review by lukasotocerny — APPROVED** [review]
LGTM now, we'll have to figure out the text because it renders on top of everything else

**Comment by huyifanx** [comment]
Yes, I am working on this problem as well. It seems that we need to change the shader of the Font

**Pull request merged**